Applying kore hardfork change (#518)
klaytn/klaytn-docs: docs/klaytn/design/computation/klaytn-virtual-machine/klaytn-virtual-machine-previous.md

### Fees Overview <a id="fees-overview"></a>

> Fees \(denominated in gas\) are charged under three distinct circumstances. Sometimes, some policies may be omitted.

Reviewer: may be omitted 한국어 원문이 어떤 뜻인가요?

Author: `몇몇 정책들은 가끔씩 생략될 수 있다.` 입니다.

Author: `some policies are ignored.` 가 더 자연스러울까요?

Reviewer 2: 아래 내용에 대해서죠? 어차피 Second, final 문장에 "may" 로 적혀있기 때문에 굳이 "Sometimes, some policies may be omitted." 써서 헷갈리게 할 필요가 있나 싶어요

Author: 네 지웠습니다

* Finally, gas may be charged due to an increase in memory usage.

Over an account's execution, the total fee payable for memory-usage payable is proportional to the smallest multiple of 32 bytes that are required to include all memory indices \(whether for read or write\) in the range. This fee is paid on a just-in-time basis; consequently, referencing an area of memory at least 32 bytes greater than any previously indexed memory will result in an additional memory usage fee. Due to this fee, it is highly unlikely that addresses will ever exceed the 32-bit bounds. That said, implementations must be able to manage this eventuality.

Storage fees have a slightly nuanced behavior. To incentivize minimization of the use of storage \(which corresponds directly to a larger state database on all nodes\), the execution fee for an operation that clears an entry from storage is not only waived but also elicits a qualified refund; in fact, this refund is effectively paid in advance because the initial usage of a storage location costs substantially more than normal usage.

#### Fee Schedule <a id="fee-schedule"></a>

> The fee schedule `G` is a tuple of 36 scalar values corresponding to the relative costs, in gas, of a number of abstract operations that a transaction may incur. Also, there's gas items to calculate the gas of the precompiled contracts called by `CALL_*` opcodes.

Reviewer: 이거 한국어 원문 좀 보여주실 수 있을까요?

Author: > The fee schedule `G` is a tuple of 36 scalar values corresponding to the relative costs, in gas, of a number of abstract operations that a transaction may incur.
> 비용표 G는 트랜잭션에서 발생할 수 있는 연산의 가스값 정보를 가지고 있는 37개 스칼라값을 가진 튜플(tuple)입니다. 
요 내용의 출처는 여기입니다. https://ko.docs.klaytn.foundation/content/klaytn/design/computation/klaytn-virtual-machine#fee-schedule

> Also, there's gas items to calculate the gas of the precompiled contracts called by `CALL_*` opcodes.
이 줄은 이번에 제가 추가한건데요, `CALL_* opcode들로부터 호출되는 precompiled contracts의 가스를 계산하는데에 쓰이는 가스 아이템들도 있다.` 라는 뜻입니다.

Precompiled contracts are special kind of contracts which usually perform complex cryptographic computations and are initiated by other contracts.

For example, gas cost can be calculated simply like below, but some gas cost calculation functions are very complex. So I would not explain the exact gas cost calculation function here.

```text
# ecrecover, sha256hash, ripemd160hash, dataCopy
Gas = XXXBaseGas + (number of words * XXXPerWordGas)

# validateSender
Gas = number of signatures * ValidateSenderGas
```

| Address | Precompiled contracts | Item | Value |
| :--- | :--- | :--- | :--- |
| 0x01 | ecrecover | EcrecoverGas | 3000 |
| 0x02 | sha256hash | Sha256BaseGas, Sha256PerWordGas | 60, 12 |
| 0x03 | ripemd160hash | Ripemd160BaseGas, Ripemd160PerWordGas | 600, 120 |
| 0x04 | dataCopy | IdentityBaseGas, IdentityPerWordGas | 15, 3 |
| 0x05 | bigModExp | ModExpQuadCoeffDiv | 20 | ​ |
| 0x06 | bn256Add | Bn256AddGas | 150 | 
| 0x07 | bn256ScalarMul | Bn256ScalarMulGas | 6000 |
| 0x08 | bn256Pairing | Bn256PairingBaseGas, Bn256PairingPerPointGas | 45000, 34000 |
| 0x09 | vmLog | VMLogBaseGas, VMLogPerByteGas | 100, 20 |
| 0x10 | feePayer | FeePayerGas | 300 |

> | 0x11 | validateSender | ValidateSenderGas | 5000 |

Reviewer 2: These addresses are changed to 0x3fd ~ 0x3ff

Reviewer 2: And, new pre-compiled contracts are introduced

Author: This page contains the klvm information used before the activation of hardforks.

Author: Btw, I think this previous page is confusing. This page was added because there was only istanbulCompatible hf at that time, but it's not effective anymore.

I'll delete this page and add the descriptions about the changes at each page like this later.
```
0x09 | blake2b(before istanbulHf, it was vmLog) |
``` 

For now, please ignore this pages.

#### Gas calculation during contract execution <a id="gas-calculation-during-contract-execution"></a>
The gas cost of one transaction is calculated through the methods described below. First, gas is added according to the transaction type and input. Then, if the contract is executed, opcodes are executed one by one until the execution ends or `STOP` operation appears. In the process, the cost is charged according to the `constantGas` defined for each opcode and the additionally defined gas calculation method. 

> Here, I will briefly explain the gas calculation logic during contract execution using the fee schedule variables defined above. As this explanation assumes a general situation, the unusual situations such as revert appears is not considered.

Reviewer: ```suggestion
Below is a brief explanation of the gas calculation logic during contract execution using the fee schedule variables defined above. As it assumes a general situation, unusual situations such as revert appears is not considered.
```
